bug: Panic due to Race Condition in HTTP Streamable Server Transport for Empty Responses

## Description

There appears to be a race condition in the HTTP Streamable Server transport when a response without a body is sent back to the client (e.g. on Initialize). The handlePost function doesn't close the done channel in this branch:
https://github.com/mark3labs/mcp-go/blob/37c4d976eb9db25ebbe7d5b6bc3103276d6b707c/server/streamable_http.go#L475-L479

Which can lead to writes to closed HTTP requests here:
https://github.com/mark3labs/mcp-go/blob/37c4d976eb9db25ebbe7d5b6bc3103276d6b707c/server/streamable_http.go#L431-L471

Here's an example panic:
```
=== RUN   TestStreamableHTTPNotificationRace
2026/03/20 13:48:06 http: superfluous response.WriteHeader call from github.com/mark3labs/mcp-go/server.(*StreamableHTTPServer).handlePost.func1.1 (streamable_http.go:456)
2026/03/20 13:48:06 ERROR: Failed to write SSE event: failed to write SSE event: http: wrote more than the declared Content-Length
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x502d72]

goroutine 52 [running]:
bufio.(*Writer).Flush(0xc0001bbcc0?)
        /usr/local/go/src/bufio/bufio.go:639 +0x12
net/http.(*chunkWriter).flush(0x416fdf?)
        /usr/local/go/src/net/http/server.go:405 +0x39
net/http.(*response).FlushError(0xc0001ec1e0)
        /usr/local/go/src/net/http/server.go:1765 +0x4b
net/http.(*response).Flush(0xb6b080?)
        /usr/local/go/src/net/http/server.go:1757 +0x13
github.com/mark3labs/mcp-go/server.(*StreamableHTTPServer).handlePost.func1.1.1()
        /go/pkg/mod/github.com/mark3labs/mcp-go@v0.45.0/server/streamable_http.go:447 +0x39
github.com/mark3labs/mcp-go/server.(*StreamableHTTPServer).handlePost.func1.1(0xc0000b4710?, 0x45b44e?, {0x8d9da0, 0xc0001ec1e0}, 0xc000014e08, {{0x83d487, 0x3}, {{0x842671, 0xb}, {0x0, ...}}}, ...)
        /go/pkg/mod/github.com/mark3labs/mcp-go@v0.45.0/server/streamable_http.go:462 +0x45b
github.com/mark3labs/mcp-go/server.(*StreamableHTTPServer).handlePost.func1()
        /go/pkg/mod/github.com/mark3labs/mcp-go@v0.45.0/server/streamable_http.go:464 +0xdd
created by github.com/mark3labs/mcp-go/server.(*StreamableHTTPServer).handlePost in goroutine 24
        /go/pkg/mod/github.com/mark3labs/mcp-go@v0.45.0/server/streamable_http.go:431 +0x80f
FAIL    bug     0.009s
```

## Code Sample

```go
// TestStreamableHTTPNotificationRace: inbound JSON-RPC notification (nil response → 202)
// plus an outbound notification on the same POST wakes the SSE pump. If done does not stop
// the pump before the response is finished, the body/writer can be reused under async writes.
// mcp-go v0.45.0 (37c4d976eb9db25ebbe7d5b6bc3103276d6b707c):
//
// Involved code (tag v0.45.0, commit 37c4d976eb9db25ebbe7d5b6bc3103276d6b707c):
//   - Notification pump goroutine:'
//     https://github.com/mark3labs/mcp-go/blob/37c4d976eb9db25ebbe7d5b6bc3103276d6b707c/server/streamable_http.go#L431-L471
//   - Nil HandleMessage → 202 path (v0.45.0 returns without close(done)):
//     https://github.com/mark3labs/mcp-go/blob/37c4d976eb9db25ebbe7d5b6bc3103276d6b707c/server/streamable_http.go#L475-L479
func TestStreamableHTTPNotificationRace(t *testing.T) {
	s := server.NewMCPServer("test-server", "1.0")
	// Inbound notification: HandleMessage returns nil → library uses the 202 Accepted path.
	// Outbound notification on the same session wakes the SSE goroutine for this POST.
	s.AddNotificationHandler("notifications/initialized", func(ctx context.Context, _ mcp.JSONRPCNotification) {
		session := server.ClientSessionFromContext(ctx)

		// Inject an outbound notification on the same POST as the inbound one: HandleMessage returns nil (202 path)
		// while the streamable-HTTP pump is still running.
		//
		// I realize this is a silly notification to send, but it's a simple way to trigger the bug.
		// I've run into this in production code when things like tool change notifications come in shortly after the initialize handler returns.
		// The notification doesn't strictly need to be in the notification handler, it just makes the timing more likely to trigger the bug.
		session.NotificationChannel() <- mcp.JSONRPCNotification{
			JSONRPC:      mcp.JSONRPC_VERSION,
			Notification: mcp.Notification{Method: "server/ping"},
		}
	})

	hs := server.NewStreamableHTTPServer(s,
		// Used to avoid setting up a session id for the test, not required to trigger the bug.
		server.WithSessionIdManager(&server.StatelessSessionIdManager{}),
	)

	ts := httptest.NewServer(hs)
	defer ts.Close()

	client := &http.Client{Transport: &http.Transport{DisableKeepAlives: true}}
	body := `{"jsonrpc":"2.0","method":"notifications/initialized"}`

	// Multiple requests make reuse races more likely.
	for i := range 200 {
		resp, err := client.Post(ts.URL+"/mcp", "application/json", strings.NewReader(body))
		if err != nil {
			t.Fatalf("iteration %d: %v", i, err)
		}
		resp.Body.Close()

		if resp.StatusCode != http.StatusAccepted {
			t.Fatalf("iteration %d: expected 202, got %d", i, resp.StatusCode)
		}
	}
}
```

## Environment

 - Go version (see `go.mod`): 1.23.0
 - mcp-go version (see `go.mod`): v0.45.0
 - Linux, amd64

## Possible Solution

I think we just need to close the `done` channel here:
https://github.com/mark3labs/mcp-go/blob/37c4d976eb9db25ebbe7d5b6bc3103276d6b707c/server/streamable_http.go#L475-L479

Becomes:
```go
	if response == nil {
		// For notifications, just send 202 Accepted with no body.
		mu.Lock()
		close(done)
		mu.Unlock()
		w.WriteHeader(http.StatusAccepted)
		return
	}
```

Happy to provide any additional information or context. I'm also happy to send a pull request if that's helpful.

--- a/server/streamable_http.go
+++ b/server/streamable_http.go
@@ -479,8 +479,14 @@ func (s *StreamableHTTPServer) handlePost(w http.ResponseWriter, r *http.Request
 	// Process message through MCPServer
 	response := s.server.HandleMessage(ctx, rawData)
 	if response == nil {
-		// For notifications, just send 202 Accepted with no body
-		w.WriteHeader(http.StatusAccepted)
+		mu.Lock()
+		close(done)
+		if !upgradedHeader {
+			mu.Unlock()
+			w.WriteHeader(http.StatusAccepted)
+		} else {
+			mu.Unlock()
+		}
 		return
 	}
 

--- a/server/streamable_http_test.go
+++ b/server/streamable_http_test.go
@@ -2840,3 +2840,36 @@ func TestStreamableHTTP_SessionIdleTTLSweeper(t *testing.T) {
 		assert.False(t, hasActiveSession, "activeSessions should be cleaned after DELETE")
 	})
 }
+
+func TestStreamableHTTPNotificationRace(t *testing.T) {
+	s := NewMCPServer("test-server", "1.0")
+	s.AddNotificationHandler("notifications/initialized", func(ctx context.Context, _ mcp.JSONRPCNotification) {
+		session := ClientSessionFromContext(ctx)
+		session.NotificationChannel() <- mcp.JSONRPCNotification{
+			JSONRPC:      mcp.JSONRPC_VERSION,
+			Notification: mcp.Notification{Method: "server/ping"},
+		}
+	})
+
+	hs := NewStreamableHTTPServer(s,
+		WithSessionIdManager(&StatelessSessionIdManager{}),
+	)
+
+	ts := httptest.NewServer(hs)
+	defer ts.Close()
+
+	client := &http.Client{Transport: &http.Transport{DisableKeepAlives: true}}
+	body := `{"jsonrpc":"2.0","method":"notifications/initialized"}`
+
+	for i := range 200 {
+		resp, err := client.Post(ts.URL+"/mcp", "application/json", strings.NewReader(body))
+		if err != nil {
+			t.Fatalf("iteration %d: %v", i, err)
+		}
+		resp.Body.Close()
+
+		if resp.StatusCode != http.StatusAccepted && resp.StatusCode != http.StatusOK {
+			t.Fatalf("iteration %d: expected 200 or 202, got %d", i, resp.StatusCode)
+		}
+	}
+}

